runtime/src/bank/builtins/core_bpf_migration/mod.rs

```diff
@@ -308,6 +310,86 @@ impl Bank {
 
         Ok(())
     }
+
+    /// Upgrade a Core BPF program.
+    /// To use this function, add a feature-gated callsite to bank's
+    /// `apply_feature_activations` function, similar to below.
+    ///
+    /// ```ignore
+    /// if new_feature_activations.contains(&feature_set::test_upgrade_program::id()) {
+    ///     self.upgrade_core_bpf_program(
+    ///        &core_bpf_program_address,
+    ///        &source_buffer_address,
+    ///        "test_upgrade_core_bpf_program",
+    ///     );
+    /// }
+    /// ```
+    #[allow(dead_code)] // Only used when an upgrade is configured.
+    pub(crate) fn upgrade_core_bpf_program(
+        &mut self,
+        core_bpf_program_address: &Pubkey,
+        source_buffer_address: &Pubkey,
+        datapoint_name: &'static str,
+    ) -> Result<(), CoreBpfMigrationError> {
+        datapoint_info!(datapoint_name, ("slot", self.slot, i64));
+
+        let target = TargetCoreBpf::new_checked(self, core_bpf_program_address)?;
+        let source = SourceBuffer::new_checked(self, source_buffer_address)?;
+
+        // Attempt serialization first before modifying the bank.
+        let new_target_program_data_account =
+            self.new_target_program_data_account(&source, target.upgrade_authority_address)?;
+
+        // Gather old and new account data sizes, for updating the bank's
+        // accounts data size delta off-chain.
+        // Since the program account is not replaced, only the program data
+        // account and the source buffer account are involved.
+        let old_data_size = checked_add(
+            target.program_data_account.data().len(),
+            source.buffer_account.data().len(),
+        )?;
+        let new_data_size = new_target_program_data_account.data().len();
+
+        // Deploy the new target Core BPF program.
+        // This step will validate the program ELF against the current runtime
+        // environment, as well as update the program cache.
+        self.directly_invoke_loader_v3_deploy(
+            &target.program_address,
+            new_target_program_data_account.data(),
+        )?;
+
+        // Calculate the lamports to burn.
+        // The target program account _and_ the target program data account
+        // will both be replaced, so burn both their lamports.
+        // The source buffer account will be cleared, so burn its lamports.
+        // The two new program accounts will need to be funded.
+        let lamports_to_burn = checked_add(
+            target.program_data_account.lamports(),
+            source.buffer_account.lamports(),
+        )?;
+        let lamports_to_fund = new_target_program_data_account.lamports();
+
+        // Update the bank's capitalization.
+        if lamports_to_burn > lamports_to_fund {
+            self.capitalization
+                .fetch_sub(checked_sub(lamports_to_burn, lamports_to_fund)?, Relaxed);
+        } else {
```

Comment:
```suggestion
        } else if lamports_to_fund > lamports_to_burn {
```
Not super critical, but we don't need to adjust the capitalization atomic if these values are equal. (Quite possible I missed suggesting this in similar capitalization adjustments.)

Reply:
Ah yeah makes sense. I think I'll update both.